### DISCUSSION TITLE
SAML Javascript Mapper for NameID

### DISCUSSION THREAD

**Discussion by VladNastase** [keycloak/keycloak#12668] [Ideas]
As part of a working infrastructure that uses Keycloak, Azure AD and LDAP (both AAD and Keycloak are synced with LDAP), we need to create (from an user attribute) the base64 string that is used as NameID in the exchange between Keycloak and AAD. We want to integrate Azure AD with Keycloak so the same user can connect to all the exposed services through Keycloak.

In our situation, using protocol mappers is the fastest and easiest method. Since the Javascript Mapper is not suitable for NameID, we developed an extension that is basically a fusion of the _Javascript Mapper_ and the _User Attribute Mapper For NameID_.

I was wondering if this is worth adding directly to Keycloak.

**Comment by darius-m** [comment]
What is the current approach? How hard would it be to create a PR that adds this functionality in Keycloak?

**Comment by belfhi** [comment]
@VladNastase could you share the code of your current approach? I think this is something that could be very useful for us! :)
edit: I found your [repository](https://github.com/VladNastase/saml-nameid-js-mapper)

**Comment by FlxPeters** [comment]
@VladNastase Can you explain why you have a demand for a new mapper?
We just made a PoC on the integration of Entra ID with Keycloak with the following built in mapper config.

![image](https://github.com/keycloak/keycloak/assets/346241/9f0aa825-f370-4246-8d15-7a6abb8ef023)

The UPN is optional I think, but the `imutableID` mapped as a NameID from Attribute does the job.
Value of `imutableID` is mapped from AD LDAP field `ObjectGUID`.

![image](https://github.com/keycloak/keycloak/assets/346241/fb71d14d-9cb4-4942-86eb-c9276eeedd16)

This results in the following token mapping:

![image](https://github.com/keycloak/keycloak/assets/346241/6a456154-d5ef-4eeb-aa84-0f840a8db928)


References:
* https://learn.microsoft.com/en-us/entra/identity/hybrid/connect/plan-connect-design-concepts#sourceanchor
* https://blog.icewolf.ch/archive/2020/07/18/how-aad-connect-calculates-immutableid/

**Reply by VladNastase** [reply]
Sorry for the late reply, it has been a while since I looked at this.

As I said in the original post, our Keycloak instance pulls users from an LDAP. The issue is that the `objectGUID` that is used to compute the `ImmutableID` is stored as a hex string on our side, and is sent as such in the SAML Assertion if we use the `User Attribute Mapper for NameID`.

For reference, here is the code we are using now, together with the extension that I put together:
```javascript
var Base64_enc = Java.type('java.util.Base64').getEncoder();

var hexToBytes = function (hex) {
    for (var bytes = [], c = 0; c < hex.length; c += 2)
        bytes.push(parseInt(hex.substr(c, 2), 16));
    return bytes;
};

var uuid = user.getFirstAttribute("ntUniqueId");
var immutableId = Base64_enc.encodeToString(hexToBytes(uuid));

exports = immutableId;
```

I'm not sure if there was an easier way to do it, but this is still working for us. It would have been nice to have it built into Keycloak so we don't have to worry an upgrade might break it in the future (though it seems unlikely). I opened a [PR](https://github.com/keycloak/keycloak/pull/14137) at the time, but it went kind of unnoticed.
